glib/src/object.rs

```diff
@@ -1981,6 +1983,14 @@ pub trait ObjectExt: ObjectType {
     /// Downgrade this object to a weak reference.
     fn downgrade(&self) -> WeakRef<Self>;
 
+    // rustdoc-stripper-ignore-next
+    /// Downgrade this object to a local weak reference.
+    /// The clousure will be called when the the object is disposed.
+    ///
+    /// Object dispose will panic if the object is disposed from the wrong thread.
```

Comment:
how would that work? `g_object_weak_ref()` isn't thread-safe

Reply:
It's thread-safe, that's not the problem. Please read the docs again :) What is not thread-safe is to use it to implement weak references. Which also brings us all the problems you have to fix here.

Reply:
> Note that the weak references created by this method are not thread-safe: they cannot safely be used in one thread if the object’s last g_object_unref() might happen in another thread. Use GWeakRef if thread-safety is required.

Honestly this isn't super clear. So this means that the notify call is thread safe. But the access to the `data` is not.

Reply:
The API itself is thread-safe (ref, unref, the callback itself). What is not thread-safe is implementing weak references with it because you have no way of "upgrading" your weak reference that is not racy with the object being destroyed. The callback is called once the object is gone, so even if you have a mutex inside the callback and inside your `upgrade()` function, you can't prevent it from "upgrading" an already destroyed object.

Reply:
makes sense. Thanks fro explaining :)